src/psyclone/psyir/frontend/fparser2.py

```
@@ -92,6 +92,141 @@ def _get_symbol_table(node):
+    :param int dim: the dimension index to check.
+    :param int index: the index of the range to check (0 is the \
+        lower bound, 1 is the upper bound).
+    :param operator: the expected operator.
+    :type operator: \
+        :py:class:`psyclone.psyir.nodes.binaryoperation.Operator.LBOUND` \
+        or :py:class:`psyclone.psyir.nodes.binaryoperation.Operator.UBOUND`
+
+    :raises NotImplementedError: if the supplied array does not have \
+        the expected properties.
+
+    '''
+    try:
+        bound = array.children[dim-1].children[index]
+        assert isinstance(bound, BinaryOperation), \
+            "Expecting BinaryOperation but found '{0}'." \
+            "".format(type(bound).__name__)
+        assert bound.operator == operator, \
+            "Expecting operator to be '{0}', but found '{1}'." \
+            "".format(str(operator), str(bound.operator))
+        reference = bound.children[0]
+        literal = bound.children[1]
+        assert isinstance(reference, Reference), \
+            "Expecting Reference but found '{0}'." \
+            "".format(type(reference).__name__)
+        assert reference.symbol is array.symbol, \
+            "Expecting Reference symbol '{0}' to be the same as array " \
+            "symbol '{1}'.".format(reference.symbol.name,
+                                   array.symbol.name)
+        assert isinstance(literal, Literal), \
+            "Expecting Literal but found '{0}'." \
+            "".format(type(literal).__name__)
+        assert literal.datatype == DataType.INTEGER, \
+            "Expecting integer datatype but found '{0}'." \
+            "".format(literal.datatype)
+        assert literal.value == str(dim), \
+            "Expecting literal value '{0}' to be the same as the " \
+            "current array dimension '{1}'.".format(literal.value,
+                                                    str(dim))
+    except AssertionError as excinfo:
+        raise NotImplementedError(
```

On looking through the tests I can see why you had misgivings about changing this to NotImplementedError. I think the problem is that this routine always raises the same exception whereas it should raise InternalError if the arguments it has been passed are invalid. I'm really not sure what it should raise if everything is valid but the bound doesn't represent a full-extent. It *feels* like we should return a `bool` to handle this.

---

OK, I've restructured to remove exceptions in the utility routines and now return True or False.

---

Thanks.